components/remote-environment-broker/internal/mapping/controller.go

```diff
@@ -157,47 +171,96 @@ func (c *Controller) processItem(key string) error {
 	// We need to check the guarantees of such solutions and choose the best one.
-	_, exists, err := c.emInformer.GetIndexer().GetByKey(key)
+	emObj, emExist, err := c.emInformer.GetIndexer().GetByKey(key)
 	if err != nil {
 		return errors.Wrapf(err, "while getting object with key %q from the store", key)
 	}
 
-	var name, namespace string
-	namespace, name, err = cache.SplitMetaNamespaceKey(key)
+	namespace, name, err := cache.SplitMetaNamespaceKey(key)
 	if err != nil {
 		return errors.Wrapf(err, "while getting name and namespace from key %q", key)
 	}
 
+	reNs, err := c.getNamespace(namespace)
+	if err != nil {
+		return err
+	}
+
+	if !emExist {
+		if err = c.ensureNsNotLabelled(reNs); err != nil {
+			return err
+		}
+		if c.manageNsBrokers {
+			if err := c.ensureNsBrokerNotRegistered(namespace); err != nil {
+				return err
+			}
+		}
+		return nil
+	}
+	if err := c.ensureNsLabelled(name, reNs); err != nil {
+		return err
+	}
+	if c.manageNsBrokers {
+		envMapping, ok := emObj.(*v1alpha1.EnvironmentMapping)
+		if !ok {
+			return errors.New("cannot cast received object to v1alpha1.EnvironmentMapping type")
+		}
+		if err := c.ensureNsBrokerRegistered(envMapping); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (c *Controller) getNamespace(namespace string) (*corev1.Namespace, error) {
 	nsObj, nsExist, nsErr := c.nsInformer.GetIndexer().GetByKey(namespace)
-	if nsErr != nil || !nsExist {
-		return errors.Wrapf(err, "cannot get the namespace: %q", namespace)
+	if nsErr != nil {
+		return nil, errors.Wrapf(nsErr, "cannot get the namespace: %q", namespace)
+	}
+
+	if !nsExist {
+		return nil, fmt.Errorf("namespace [%s] not found", namespace)
 	}
 
 	reNs, ok := nsObj.(*corev1.Namespace)
 	if !ok {
-		return errors.New("cannot cast received object to corev1.Namespace type")
+		return nil, errors.New("cannot cast received object to corev1.Namespace type")
 	}
+	return reNs, nil
+}
 
-	if !exists {
-		if err = c.deleteNsAccLabel(reNs); err != nil {
-			return errors.Wrapf(err, "cannot delete AccessLabel from the namespace: %q", namespace)
-		}
+func (c *Controller) ensureNsBrokerRegistered(envMapping *v1alpha1.EnvironmentMapping) error {
+	brokerExist, err := c.nsBrokerFacade.Exist(envMapping.Namespace)
+	if err != nil {
+		return errors.Wrapf(err, "while checking if namespaced broker exist in namespace [%s]", envMapping.Namespace)
+	}
+	if brokerExist {
 		return nil
 	}
-	var label string
-	label, err = c.getReAccLabel(name)
-	if err != nil {
-		return errors.Wrapf(err, "cannot get AccessLabel from RE: %q", name)
+	if err := c.nsBrokerFacade.Create(envMapping.Namespace, c.systemNamesapce); err != nil {
+		return errors.Wrapf(err, "while creating namespaced broker in namespace [%s]", envMapping.Namespace)
 	}
-	err = c.applyNsAccLabel(reNs, label)
+
+	return nil
+}
+
+func (c *Controller) ensureNsBrokerNotRegistered(namespace string) error {
+	brokerExist, err := c.nsBrokerFacade.Exist(namespace)
```

Comment:
why do we check if broker exists? Let's remove the check. The brokerFacade.Create skips error AlreadyExists.
Current implementation may be error prone - imagine service was not created (k8s API was not available for the service creation). The implementation of the Facade goes to the next operation - create Broker. At the end - broker is created, service - not. Error will be returned. The controller operation will be retried, but broker exists, so nothing will be done.
other solution - add checking if service exists in the Exists method

Reply:
Exist now will check additionally if Service exist